src/network/network_gui.cpp

```diff
@@ -1806,166 +1793,543 @@ static void PopupClientList(ClientID client_id, int x, int y)
+	}
+
+	Button(StringID label, StringID tooltip, Colours colour, ClientID client_id, ButtonClientCallback proc) :
+		label(label),
+		tooltip(tooltip),
+		colour(colour),
+		company_proc(nullptr),
+		client_id(client_id),
+		client_proc(proc)
+	{
+		Dimension d = GetStringBoundingBox(label);
+		this->height = d.height;
+		this->width = d.width + WD_FRAMERECT_LEFT + WD_FRAMERECT_RIGHT;
+	}
+};
+
 /**
  * Main handle for clientlist
  */
 struct NetworkClientListWindow : Window {
-	int selected_item;
+private:
+	ClientListWidgets query_widget;
+	CompanyID join_company;
 
-	uint server_client_width;
+	Scrollbar *vscroll;
 	uint line_height;
+	uint line_count;
 
-	Dimension icon_size;
+	std::map<uint, std::vector<Button>> buttons;
 
-	NetworkClientListWindow(WindowDesc *desc, WindowNumber window_number) :
-			Window(desc),
-			selected_item(-1)
+	/**
+	 * Chat button on a Company is clicked.
+	 */
+	static void OnClickCompanyChat(NetworkClientListWindow *w, Point pt, CompanyID company_id)
 	{
-		this->InitNested(window_number);
+		ShowNetworkChatQueryWindow(DESTTYPE_TEAM, company_id);
 	}
 
 	/**
-	 * Finds the amount of clients and set the height correct
+	 * Join button on a Company is clicked.
 	 */
-	bool CheckClientListHeight()
+	static void OnClickCompanyJoin(NetworkClientListWindow *w, Point pt, CompanyID company_id)
+	{
+		if (_network_server) {
+			NetworkServerDoMove(CLIENT_ID_SERVER, company_id);
+			MarkWholeScreenDirty();
+		} else if (NetworkCompanyIsPassworded(company_id)) {
+			w->query_widget = WID_CL_COMPANY_JOIN;
+			w->join_company = company_id;
+			ShowQueryString(STR_EMPTY, STR_NETWORK_NEED_COMPANY_PASSWORD_CAPTION, NETWORK_PASSWORD_LENGTH, w, CS_ALPHANUMERAL, QSF_PASSWORD);
+		} else {
+			NetworkClientRequestMove(company_id);
+		}
+	}
+
+	/**
+	 * Admin button on a Client is clicked.
+	 */
+	static void OnClickClientAdmin(NetworkClientListWindow *w, Point pt, ClientID client_id)
 	{
-		int num = 0;
+		PopupClientList(client_id, pt.x + w->left, pt.y + w->top);
+	}
 
-		/* Should be replaced with a loop through all clients */
-		for (const NetworkClientInfo *ci : NetworkClientInfo::Iterate()) {
-			if (ci->client_playas != COMPANY_INACTIVE_CLIENT) num++;
+	/**
+	 * Chat button on a Client is clicked.
+	 */
+	static void OnClickClientChat(NetworkClientListWindow *w, Point pt, ClientID client_id)
+	{
+		ShowNetworkChatQueryWindow(DESTTYPE_CLIENT, client_id);
+	}
+
+	/**
+	 * Rebuild the list, meaning: calculate the lines needed and what buttons go on which line.
+	 */
+	void RebuildList()
+	{
+		NetworkClientInfo *own_ci = NetworkClientInfo::GetByClientID(_network_own_client_id);
+
+		this->buttons.clear();
+		this->line_count = 0;
+
+		/* Companies */
+		for (const Company *c : Company::Iterate()) {
+			this->buttons[line_count].push_back(Button(STR_NETWORK_CLIENT_LIST_CHAT, STR_NETWORK_CLIENT_LIST_CHAT_COMPANY_TOOLTIP, COLOUR_ORANGE, c->index, &this->OnClickCompanyChat));
```

Comment:
I think you can use `emplace_back(` instead of `push_back(Button(`. The former will call the constructor directly on the memory location in the vector, instead of first making the Button object and then copying/moving that into the vector.